components/tor/tor_file_watcher.cc

```diff
@@ -0,0 +1,336 @@
+//      port and mtime if successful; return false on failure.
+//
+bool TorFileWatcher::EatControlPort(int& port, base::Time& mtime) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(watch_sequence_checker_);
+  DCHECK(polling_);
+
+  // Open the control port file.
+  base::FilePath portpath = watch_dir_path_.AppendASCII(kControlPortName);
+  base::File portfile(portpath, base::File::FLAG_OPEN | base::File::FLAG_READ);
+  if (!portfile.IsValid()) {
+    VLOG(0) << "tor: failed to open control port";
+    return false;
+  }
+
+  // Get the file's info, including modification time.
+  base::File::Info info;
+  if (!portfile.GetInfo(&info)) {
+    VLOG(0) << "tor: failed to stat control port";
+    return false;
+  }
+
+  // Read up to 27/28 octets, the maximum we will ever need.
+  const size_t kBufSiz = strlen(kControlPortMaxTmpl);
+  char buf[kBufSiz];
+  int nread = portfile.ReadAtCurrentPos(buf, sizeof buf);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read control port";
+    return false;
+  }
+  if (static_cast<size_t>(nread) < strlen(kControlPortMinTmpl)) {
+    VLOG(0) << "tor: control port truncated";
+    return false;
+  }
+  DCHECK(static_cast<size_t>(nread) <= sizeof buf);
+
+  std::string text(buf, 0, nread);
+
+  // Sanity-check the content.
+  if (!base::StartsWith(text, "PORT=", base::CompareCase::SENSITIVE) ||
+      !base::EndsWith(text, kLineBreak, base::CompareCase::SENSITIVE)) {
+    VLOG(0) << "tor: invalid control port: "
+            << "`" << text << ";";  // XXX escape
+    return false;
+  }
+
+  // Verify that it's localhost.
+  const char expected[] = "PORT=127.0.0.1:";
+  if (!base::StartsWith(text, expected, base::CompareCase::SENSITIVE)) {
+    VLOG(0) << "tor: control port has non-local control address";
+    return false;
+  }
+
+  // Parse it!
+  std::string portstr(text, strlen(expected),
+                      nread - strlen(kLineBreak) - strlen(expected));
+  if (!base::StringToInt(portstr, &port)) {
+    VLOG(0) << "tor: failed to parse control port: "
+            << "`" << portstr << "'";  // XXX escape
+    return false;
+  }
+  if (port > 65535) {
+    VLOG(0) << "tor: port overflow";
+    return false;
+  }
+  mtime = info.last_modified;
+  VLOG(3) << "Control port " << port << ", mtime " << mtime;
+  return true;
+}
+
+bool TorFileWatcher::EatOldPid(base::ProcessId* id) {
+  DCHECK_CALLED_ON_VALID_SEQUENCE(watch_sequence_checker_);
+  DCHECK(id);
+
+  // Open the tor pid file.
+  base::FilePath pidpath = watch_dir_path_.AppendASCII(kTorPidName);
+  base::File pidfile(pidpath, base::File::FLAG_OPEN | base::File::FLAG_READ);
+  if (!pidfile.IsValid()) {
+    VLOG(0) << "tor: failed to open tor.pid";
+    return false;
+  }
+
+  const size_t kBufSiz = pidfile.GetLength();
+  char buf[kBufSiz];
+  int nread = pidfile.ReadAtCurrentPos(buf, sizeof buf);
+  if (nread < 0) {
+    VLOG(0) << "tor: failed to read tor pid file";
+    return false;
+  }
+
+  std::string pid(buf, 0, nread - strlen(kLineBreak));
```

Comment:
it looks like overflow if `nread == 1` and `strlen(kLineBreak) == 2`. Or `0` and `1`

Reply:
The whole logic of `CheckingOldTorProcess` and `EatOldPid` will go away with
https://github.com/brave/brave-browser/issues/14044
I can do it in this PR if you want

Reply:
fixed in https://github.com/brave/brave-core/pull/7927/commits/4210219b125b5cee9dd6425dc5b7af215e7f67b7